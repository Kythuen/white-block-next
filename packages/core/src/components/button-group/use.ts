import { defineComponent, h } from 'vue'
import ButtonGroup from '@/components/button-group/button-group.vue'
import { getRawSlots } from '@/utils'
import type { Props } from '@/components/button-group/config'

export function useButtonGroup(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () =>
        h(ButtonGroup, { ...(options || {}) } as any, getRawSlots(slots))
    }
  })
  return { node }
}
