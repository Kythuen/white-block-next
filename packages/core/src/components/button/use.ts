import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import Button from './button.vue'

export function useButton(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(Button, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
