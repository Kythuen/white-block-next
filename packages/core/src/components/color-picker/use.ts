import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import ColorPicker from './color-picker.vue'

export function useColorPicker(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(ColorPicker, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
