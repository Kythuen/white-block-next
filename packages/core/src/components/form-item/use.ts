import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import FormItem from './form-item.vue'

export function useFormItem(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(FormItem, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
