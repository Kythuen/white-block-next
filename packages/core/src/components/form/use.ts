import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import Form from './form.vue'

export function useForm(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(Form, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
