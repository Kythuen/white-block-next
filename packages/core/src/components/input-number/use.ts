import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import InputNumber from './input-number.vue'

export function useInputNumber(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(InputNumber, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
