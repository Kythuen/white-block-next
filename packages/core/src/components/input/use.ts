import { SizeProps, DisabledProps, ValueProps } from '@/components/api'
import { Props } from '@/components/input/api'
import Input from '@/components/input/input.vue'
import { getRawSlots } from '@/utils'
import { defineComponent, h, ref } from 'vue'

export function useInput(
  options?: Partial<Props & SizeProps & ValueProps<string> & DisabledProps>
) {
  const inputValue = ref('')
  const node = defineComponent({
    setup(_, { slots }) {
      return () =>
        h(
          Input,
          {
            'modelValue': inputValue.value,
            'onUpdate:modelValue': (val: string) => {
              inputValue.value = val
            },
            ...(options || {})
          },
          getRawSlots(slots)
        )
    }
  })
  return { node, value: inputValue }
}
