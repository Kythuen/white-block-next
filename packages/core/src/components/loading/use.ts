import { defineComponent, h } from 'vue'
import { getRawSlots } from '../../utils'
import { Props } from './config'
import Loading from './loading.vue'

export function useLoading(options?: Partial<Props>) {
  const node = defineComponent({
    setup(_, { slots }) {
      return () => h(Loading, { ...(options || {}) }, getRawSlots(slots))
    }
  })
  return { node }
}
