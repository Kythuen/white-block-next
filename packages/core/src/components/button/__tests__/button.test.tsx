import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import Button from '../button.vue'
import { useButton } from '../use'

describe('# Button', () => {
  describe('## Base', () => {
    it('### default config', () => {
      const wrapper = mount(Button)
      expect(wrapper.classes()).toEqual([
        'wb-button',
        'wb-button--base',
        'wb-button--primary',
        'wb-button--rectangle',
        'wb-button--md'
      ])
      expect(wrapper.props()).toEqual({
        block: false,
        color: '',
        content: '',
        default: undefined,
        disabled: false,
        loading: false,
        onClick: undefined,
        prefix: undefined,
        shape: 'rectangle',
        size: 'md',
        suffix: undefined,
        tag: 'button',
        theme: 'primary',
        type: 'base'
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('## Slots', () => {
    describe('### content', () => {
      it('#### slot: content', () => {
        const wrapper = mount(Button, {
          slots: { content: 'Button' }
        })
        expect(wrapper.text()).toEqual('Button')
      })
      it('#### prop: content', () => {
        const wrapper = mount(Button, {
          props: { content: 'Button' }
        })
        expect(wrapper.text()).toEqual('Button')
      })
      it('#### prop: content[node]', () => {
        const wrapper = mount(Button, {
          slots: { content: () => <span>Button</span> }
        })
        expect(wrapper.text()).toEqual('Button')
      })
      it('#### prop: content[node] function', () => {
        const wrapper = mount(Button, {
          props: { content: () => <span>Button</span> }
        })
        expect(wrapper.text()).toEqual('Button')
      })
    })

    describe('### default', () => {
      it('#### slot: default', () => {
        const wrapper = mount(Button, {
          slots: { default: () => <span>Button</span> }
        })
        expect(wrapper.text()).toEqual('Button')
      })
      it('#### prop: default', () => {
        const wrapper = mount(Button, {
          props: { default: 'Button' }
        })
        expect(wrapper.text()).toEqual('Button')
      })
      it('#### prop: default[node]', () => {
        const wrapper = mount(Button, {
          slots: { default: () => <span>Button</span> }
        })
        expect(wrapper.text()).toEqual('Button')
      })
    })

    describe('### prefix', () => {
      it('#### slot: prefix', () => {
        const wrapper = mount(Button, {
          slots: { prefix: 'Prefix', content: 'Button' }
        })
        expect(wrapper.text()).toEqual('PrefixButton')
      })
      it('#### prop: prefix[node]', () => {
        const wrapper = mount(Button, {
          props: { prefix: <span>Prefix</span>, content: 'Button' }
        })
        expect(wrapper.text()).toEqual('PrefixButton')
      })
    })

    describe('### suffix', () => {
      it('#### slot: suffix', () => {
        const wrapper = mount(Button, {
          slots: { suffix: 'Suffix', content: 'Button' }
        })
        expect(wrapper.text()).toEqual('ButtonSuffix')
      })
      it('#### property: suffix[node]', () => {
        const wrapper = mount(Button, {
          props: { suffix: <span>Suffix</span>, content: 'Button' }
        })
        expect(wrapper.text()).toEqual('ButtonSuffix')
      })
    })
  })

  describe('## Events', () => {
    describe('### click', () => {
      it('#### click', async () => {
        const wrapper = mount(Button)
        await wrapper.trigger('click')
        expect(wrapper.emitted()).toHaveProperty('click')
      })
      it('#### click: disabled', () => {
        const wrapper = mount(Button, { disabled: true })
        expect(wrapper.emitted()).toEqual({})
      })
      it('#### click: loading', () => {
        const wrapper = mount(Button, { loading: true })
        expect(wrapper.emitted()).toEqual({})
      })
    })
  })

  describe('## Props', () => {
    it('### size', () => {
      const wrapper = mount(Button, {
        props: { size: 'sm' }
      })
      expect(wrapper.classes()).toContain('wb-button--sm')
    })
    it('### size:number', () => {
      const wrapper = mount(Button, {
        props: { size: 32 }
      })
      const elStyle = wrapper.find('button').attributes('style')
      expect(
        elStyle?.includes('--wb-button-height: calc(32 / 16 * 1rem)')
      ).toBe(true)
    })

    it('### type', () => {
      const wrapper = mount(Button, {
        props: { type: 'plain' }
      })
      expect(wrapper.classes()).toContain('wb-button--plain')
    })

    it('### theme', () => {
      const wrapper = mount(Button, {
        props: { theme: 'default' }
      })
      expect(wrapper.classes()).toContain('wb-button--default')
    })

    it('### shape', () => {
      const wrapper = mount(Button, {
        props: { shape: 'circle' }
      })
      expect(wrapper.classes()).toContain('wb-button--circle')
    })

    it('### color', () => {
      const wrapper = mount(Button, {
        props: { color: 'rgb(255 165 0)' }
      })
      const elStyle = wrapper.find('button').attributes('style')
      expect(
        elStyle?.includes('--wb-button-color-main: rgb(255 165 0 / 1);')
      ).toBe(true)
    })

    it('### tag', () => {
      const wrapper = mount(Button, {
        props: { tag: 'a' }
      })
      expect(wrapper.element.tagName).toBe('A')
    })

    it('### disabled', () => {
      const wrapper = mount(Button, {
        props: { disabled: true }
      })
      expect(wrapper.classes()).toContain('wb-button--disabled')
    })

    it('### loading', () => {
      const wrapper = mount(Button, {
        props: { loading: true }
      })
      expect(wrapper.classes()).toContain('wb-button--loading')
    })
  })

  describe('## Composables', () => {
    const { node: MyButton } = useButton()
    it('### no configs', () => {
      const wrapper = mount(MyButton)
      expect(wrapper.classes()).toContain('wb-button--md')
    })
    const { node: MyButton2 } = useButton({
      content: 'button',
      size: 'sm'
    })
    it('### with props', async () => {
      const wrapper = mount(MyButton2, { props: { block: true } })
      expect(wrapper.classes()).toEqual(
        expect.arrayContaining(['wb-button--sm', 'wb-button--block'])
      )
      await wrapper.trigger('click')
      expect(wrapper.emitted()).toHaveProperty('click')
    })
  })
})
