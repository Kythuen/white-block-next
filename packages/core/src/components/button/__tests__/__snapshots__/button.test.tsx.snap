// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`# Button > ## Base > ### default config 1`] = `
<button
  aria-disabled="false"
  bg="$wb-button-background hover:$wb-button-background-hover active:$wb-button-background-active"
  box="border"
  class="wb-button wb-button--base wb-button--primary wb-button--rectangle wb-button--md"
  color="$wb-button-text hover:$wb-button-text-hover active:$wb-button-text-active"
  cursor="pointer"
  flex="inline row nowrap"
  focus-visible="outline outline-2 outline-offset-2 outline-$wb-color-primary"
  gap="$wb-button-gap"
  h="$wb-button-height"
  items="center"
  justify="center"
  outline="none"
  p="x-$wb-button-padding"
  ring="1 inset $wb-button-border hover:$wb-button-border-hover active:$wb-button-border-active"
  rounded="$wb-button-radius"
  select="none"
  transition="all ease duration-200"
  type="button"
>
  <!--v-if-->
  <!---->
  <!---->
</button>
`;
